File: python/day20.py
import math
from functools import reduce
import operator

OPPOSITE_DIR = [2, 3, 0, 1]
OFFS = [(0, -1), (1, 0), (0, 1), (-1, 0)]
N = 10


def parse_tile(data):
    lines = [line.strip() for line in data.split('\n')]
    try:
        tile_id = int(lines[0].split()[1][:-1])
        return (tile_id, lines[1:])
    except:
        return None


def get_bit_mask(seq):
    res = 0
    for c in seq:
        res = (res << 1) | (c == '#')
    return res


def reverse_bits(val):
    res = 0
    for i in range(N):
        res = res << 1 | (val & 1)
        val = val >> 1
    return res


def get_borders(tile):
    (tile_id, lines) = tile
    top = get_bit_mask(lines[0])
    bottom = get_bit_mask(lines[-1])
    left = get_bit_mask(line[0] for line in lines)
    right = get_bit_mask(line[-1] for line in lines)
    return (top, right, bottom, left)


def rotate(borders):
    (t, r, b, l) = borders
    return (reverse_bits(l), t, reverse_bits(r), b)


def get_transforms(borders):
    for i in range(4):
        yield borders
        (t, r, b, l) = borders
        # flip horizontal/vertical
        yield (reverse_bits(t), l, reverse_bits(b), r)
        yield (b, reverse_bits(r), t, reverse_bits(l))
        borders = rotate(borders)
    pass


def create_mappings(tiles):
    mappings = [{}, {}, {}, {}]
    for (i, tile) in enumerate(tiles):
        borders = get_borders(tile)
        for tm in set(get_transforms(borders)):
            for side in range(4):
                mask = tm[side]
                if mask not in mappings[side]:
                    mappings[side][mask] = []
                mappings[side][mask].append((i, tm))
    return mappings


def find_corners(tiles, mapping):
    res = []
    for (tile_idx, t) in enumerate(tiles):
        borders = get_borders(t)
        k = 0
        for i, mask in enumerate(borders):
            m = set(x[0] for x in mapping[OPPOSITE_DIR[i]][mask])
            if len(m) == 1 and tile_idx in m:
                k += 1
        if k == 2:
            res.append(tile_idx)
    return res


def get_layout(tiles):
    mappings = create_mappings(tiles)
    n = len(tiles)

    used = [False] * n
    num_used = 0
    res = {}

    def layout(tile_idx, orientation, x, y):
        nonlocal num_used
        if used[tile_idx]:
            return False

        for (i, mask) in enumerate(orientation):
            x1, y1 = x + OFFS[i][0], y + OFFS[i][1]
            if (x1, y1) in res and res[(x1, y1)][1][OPPOSITE_DIR[i]] != mask:
                return False

        used[tile_idx] = True
        res[(x, y)] = (tile_idx, orientation)
        num_used += 1
        if num_used == n:
            return True

        found_fit = False
        for i, mask in enumerate(orientation):
            x1, y1 = x + OFFS[i][0], y + OFFS[i][1]
            if (x1, y1) in res:
                continue
            for (ci, co) in mappings[OPPOSITE_DIR[i]][mask]:
                if layout(ci, co, x1, y1):
                    found_fit = True
                    break
            if found_fit:
                break
        if not found_fit:
            used[tile_idx] = False
            num_used -= 1
            res.pop((x, y), None)
        return found_fit

    corners = find_corners(tiles, mappings)
    start_idx = corners[0]
    layout(start_idx, get_borders(tiles[start_idx]), 0, 0)
    return res


def build_image(tiles, layout):
    minx = min(x[0] for x in layout.keys())
    miny = min(x[1] for x in layout.keys())
    maxx = max(x[0] for x in layout.keys())
    maxy = max(x[1] for x in layout.keys())
    res = [['.' for j in range((maxx - minx + 1) * (N - 2))]
           for i in range((maxy - miny + 1) * (N - 2))]
    for pos, (tile_idx, b) in layout.items():
        borders = list(get_transforms(get_borders(tiles[tile_idx])))
        tm = borders.index(b)
        ang = tm // 3
        flip_x = (tm % 3) == 1
        flip_y = (tm % 3) == 2
        x = (pos[0] - minx) * (N - 2)
        y = (pos[1] - miny) * (N - 2)
        img = tiles[tile_idx][1]
        for i in range(N - 2):
            for j in range(N - 2):
                c = img[j][i]
                cx, cy = i + 1, j + 1
                if flip_x:
                    cx = N - cx - 1
                if flip_y:
                    cy = N - cy - 1

                if ang == 1:
                    cx, cy = cy, N - cx - 1
                elif ang == 2:
                    cx, cy = N - cx - 1, N - cy - 1
                elif ang == 3:
                    cx, cy = N - cy - 1, cx

                res[y + j][x + i] = img[cy][cx]
    return [''.join(s) for s in res]


def count_hashes(image, pattern):
    w, h = len(image[0]), len(image)

    def matches_sub(x, y):
        checked = 0
        for j, p in enumerate(pattern):
            for i, c in enumerate(p):
                x1, y1 = x + i, y + j
                if x1 < 0 or y1 < 0 or x1 >= w or y1 >= h:
                    continue
                if pattern[j][i] == '#' and image[y1][x1] != '#':
                    return False
                checked += 1
        return checked > 0
    res = [list(line) for line in image]
    for y in range(h - len(pattern) + 1):
        for x in range(w - len(pattern[0]) + 1):
            if matches_sub(x, y):
                for j, p in enumerate(pattern):
                    for i, c in enumerate(p):
                        x1, y1 = x + i, y + j
                        if x1 < 0 or y1 < 0 or x1 >= w or y1 >= h:
                            continue
                        if pattern[j][i] == '#':
                            res[y + j][x + i] = 'O'
    num_o, num_x = 0, 0
    for y in res:
        for c in y:
            num_o += c == 'O'
            num_x += c == '#'

    return (num_o, num_x)


def rotate_img(image):
    res = [list(line) for line in image]
    n = len(image)
    for y in range(n):
        for x in range(n):
            res[y][x] = image[x][n - y - 1]
    return [''.join(s) for s in res]


def fliph(image):
    return [line[::-1] for line in image]


def flipw(image):
    return [line for line in image[::-1]]


def get_img_transforms(image):
    for i in range(4):
        yield image
        yield fliph(image)
        yield flipw(image)
        image = rotate_img(image)


def compute_roughness(tiles, pattern):
    layout = get_layout(tiles)
    image = build_image(tiles, layout)
    for img in get_img_transforms(image):
        (co, ch) = count_hashes(img, pattern)
        if co != 0:
            return ch


def solution():
    chunks = open('../data/20.txt', 'r').read().split('\n\n')
    tiles = [t for t in (parse_tile(chunk) for chunk in chunks) if t]
    res1 = reduce(operator.mul, (tiles[t][0]
                                 for t in find_corners(tiles, create_mappings(tiles))), 1)
    print(f'Answer 1: {res1}')

    pattern = open('../data/20.pattern.txt', 'r').read().split('\n')
    res2 = compute_roughness(tiles, pattern)
    print(f'Answer 2: {res2}')

File: rust/src/day20.rs
use crate::common;
use std::collections::HashMap;
use std::fmt;
use std::fs;

type TileId = u64;

#[derive(Debug)]
struct Tile {
    pt: Vec<Vec<char>>,
    id: TileId,
}

#[derive(Eq, PartialEq, Hash, Clone, Debug)]
enum Dir {
    N,
    E,
    S,
    W,
}

impl Dir {
    fn opposite(&self) -> Dir {
        match self {
            Dir::N => Dir::S,
            Dir::E => Dir::W,
            Dir::S => Dir::N,
            Dir::W => Dir::E,
        }
    }
    fn opposite_idx(&self) -> usize {
        Dir::ENUM
            .iter()
            .position(|d| *d == self.opposite())
            .unwrap()
    }
    const ENUM: [Dir; 4] = [Dir::N, Dir::E, Dir::S, Dir::W];
    const OFFS: [(i64, i64); 4] = [(0, -1), (1, 0), (0, 1), (-1, 0)];
}

#[derive(Debug, Clone)]
struct Transform {
    rotation: Dir,
    flip_x: bool,
    flip_y: bool,
}

type TileSet = Vec<Tile>;

#[derive(Debug, Clone)]
struct TileInstance {
    tile_idx: usize,
    tm: Transform,
    edges: [String; 4],
}

#[derive(Eq, PartialEq, Hash, Clone, Debug)]
struct TileMappingKey {
    dir: Dir,
    edge: String,
}

type TileMapping = HashMap<TileMappingKey, Vec<TileInstance>>;
type TileLayout = HashMap<(i64, i64), TileInstance>;

impl Tile {
    fn parse(text: &str, has_header: bool) -> Option<Tile> {
        let lines: Vec<&str> = text.split("\n").map(|s| s.trim()).collect();
        if lines.len() < 2 {
            return None;
        }
        let (id, rows) = if has_header {
            let header = lines[0].split(" ").collect::<Vec<&str>>()[1];
            (
                header[..header.len() - 1].parse().ok()?,
                lines[1..].to_vec(),
            )
        } else {
            (0, lines[0..].to_vec())
        };
        Some(Tile {
            id: id,
            pt: rows.iter().map(|s| s.chars().collect()).collect(),
        })
    }
    fn empty(w: usize, h: usize) -> Tile {
        Tile {
            id: 0,
            pt: vec![vec!['.'; w]; h],
        }
    }
    fn count(&self, ch: char) -> usize {
        self.pt
            .iter()
            .flat_map(|row| row.iter().map(|c| (*c == ch) as usize))
            .sum()
    }
    fn get_extents(&self) -> (usize, usize) {
        (self.pt[0].len(), self.pt.len())
    }
    fn get_edge(&self, dir: &Dir) -> String {
        match dir {
            Dir::N => self.pt[0].iter().collect(),
            Dir::E => self.pt.iter().map(|s| *s.last().unwrap()).collect(),
            Dir::S => self.pt.last().unwrap().iter().collect(),
            Dir::W => self.pt.iter().map(|s| s[0]).collect(),
        }
    }
    fn get_edges(&self) -> [String; 4] {
        [
            self.get_edge(&Dir::N),
            self.get_edge(&Dir::E),
            self.get_edge(&Dir::S),
            self.get_edge(&Dir::W),
        ]
    }
    fn transform(&self, tm: &Transform) -> Tile {
        let n = self.pt.len();
        let mut res = Tile::empty(n, n);
        if n == 0 {
            return res;
        }
        assert!(n == self.pt[0].len());
        for j in 0..n {
            for i in 0..n {
                let (mut x, mut y) = (i, j);
                if tm.flip_x {
                    x = n - x - 1;
                }
                if tm.flip_y {
                    y = n - y - 1;
                }
                let (rx, ry) = match tm.rotation {
                    Dir::N => (x, y),
                    Dir::E => (y, n - x - 1),
                    Dir::S => (n - x - 1, n - y - 1),
                    Dir::W => (n - y - 1, x),
                };
                res.pt[j][i] = self.pt[ry][rx];
            }
        }
        res
    }
    fn blit(&mut self, tile: &Tile, x: usize, y: usize, border: usize) {
        let n = tile.pt.len();
        for j in 0..(n - border * 2) {
            for i in 0..(n - border * 2) {
                self.pt[y + j][x + i] = tile.pt[j + border][i + border];
            }
        }
    }
    fn matches(&self, tile: &Tile, x: usize, y: usize) -> bool {
        for (j, row) in tile.pt.iter().enumerate() {
            for (i, c) in row.iter().enumerate() {
                if *c == '#' && self.pt[y + j][x + i] != *c {
                    return false;
                }
            }
        }
        true
    }
    fn find_matches(&self, pattern: &Tile) -> Vec<(i64, i64)> {
        let mut res = Vec::new();
        let (w, h) = self.get_extents();
        let (tw, th) = pattern.get_extents();
        for j in 0..(h - th + 1) {
            for i in 0..(w - tw + 1) {
                if self.matches(pattern, i, j) {
                    res.push((i as i64, j as i64));
                }
            }
        }
        res
    }
}

impl fmt::Display for Tile {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        for row in &self.pt {
            for c in row.iter() {
                write!(f, "{}", c)?;
            }
            write!(f, "\n")?;
        }
        Ok(())
    }
}

fn create_mapping(tiles: &TileSet) -> TileMapping {
    let mut res = TileMapping::new();
    for (i, tile) in tiles.iter().enumerate() {
        for flip_x in &[false, true] {
            for flip_y in &[false, true] {
                if *flip_x && *flip_y {
                    continue;
                }
                for rotation in &Dir::ENUM {
                    let tm = Transform {
                        rotation: rotation.clone(),
                        flip_x: *flip_x,
                        flip_y: *flip_y,
                    };
                    let tile_transformed = tile.transform(&tm);
                    for dir in &Dir::ENUM {
                        let edge = tile_transformed.get_edge(dir);
                        let key = TileMappingKey {
                            dir: dir.clone(),
                            edge: edge,
                        };
                        let tile_instance = TileInstance {
                            tile_idx: i,
                            tm: tm.clone(),
                            edges: tile_transformed.get_edges(),
                        };
                        res.entry(key).or_insert(Vec::new()).push(tile_instance);
                    }
                }
            }
        }
    }
    res
}

fn find_corners(tiles: &TileSet, mapping: &TileMapping) -> Vec<usize> {
    let mut res: Vec<usize> = Vec::new();
    for (i, tile) in tiles.iter().enumerate() {
        let mut num_unique_edges = 0;
        for dir in &Dir::ENUM {
            let key = TileMappingKey {
                dir: dir.opposite(),
                edge: tile.get_edge(dir),
            };
            num_unique_edges += (match mapping.get(&key) {
                Some(matches) => matches.iter().map(|t| (i != t.tile_idx) as u32).sum(),
                _ => 0,
            } == 0) as u32;
        }
        if num_unique_edges == 2 {
            res.push(i);
        }
    }
    res
}

fn get_layout(tiles: &TileSet, mapping: &TileMapping) -> TileLayout {
    fn rec(
        tile_idx: usize,
        tm: &Transform,
        pos: (i64, i64),
        tiles: &TileSet,
        mapping: &TileMapping,
        used: &mut Vec<bool>,
        res: &mut TileLayout,
    ) -> bool {
        if used[tile_idx] {
            return false;
        }
        // find if existing neighbors fit with the transformed tile's edges
        let tile_transformed = tiles[tile_idx].transform(&tm);
        let edges = tile_transformed.get_edges();
        for (i, edge) in edges.iter().enumerate() {
            let offs = Dir::OFFS[i];
            let pos1 = (pos.0 + offs.0, pos.1 + offs.1);
            match res.get(&pos1) {
                Some(t) => {
                    if t.edges[Dir::ENUM[i].opposite_idx()] != *edge {
                        return false;
                    }
                }
                _ => (),
            }
        }
        used[tile_idx] = true;
        res.insert(
            pos,
            TileInstance {
                tile_idx: tile_idx,
                tm: tm.clone(),
                edges: edges.clone(),
            },
        );
        if res.len() == tiles.len() {
            return true;
        }
        // recurse into neighbors
        let mut found_fit = false;
        for (i, edge) in edges.iter().enumerate() {
            let offs = Dir::OFFS[i];
            let pos1 = (pos.0 + offs.0, pos.1 + offs.1);
            if res.contains_key(&pos1) {
                // neighbor position is already occupied
                continue;
            }
            let key = TileMappingKey {
                dir: Dir::ENUM[i].opposite(),
                edge: edge.clone(),
            };
            match mapping.get(&key) {
                Some(instances) => {
                    if instances
                        .iter()
                        .any(|inst| rec(inst.tile_idx, &inst.tm, pos1, tiles, mapping, used, res))
                    {
                        found_fit = true;
                        break;
                    }
                }
                _ => (),
            }
        }
        if !found_fit {
            used[tile_idx] = false;
            res.remove(&pos);
        }
        found_fit
    }
    let mut res = TileLayout::new();
    let mut used = vec![false; tiles.len()];
    let corners = find_corners(tiles, mapping);
    let start_idx = corners[0];
    rec(
        start_idx,
        &Transform {
            rotation: Dir::N,
            flip_x: false,
            flip_y: false,
        },
        (0, 0),
        tiles,
        mapping,
        &mut used,
        &mut res,
    );
    res
}

fn blit_layout(tiles: &TileSet, layout: &TileLayout) -> Tile {
    let minx = layout.keys().map(|x| x.0).min().unwrap();
    let maxx = layout.keys().map(|x| x.0).max().unwrap();
    let miny = layout.keys().map(|x| x.1).min().unwrap();
    let maxy = layout.keys().map(|x| x.1).max().unwrap();
    let n = tiles[0].pt.len() - 2;
    let mut res = Tile::empty(
        n * (maxx - minx + 1) as usize,
        n * (maxy - miny + 1) as usize,
    );
    for (pos, instance) in layout {
        let tile_transformed = tiles[instance.tile_idx].transform(&instance.tm);
        res.blit(
            &tile_transformed,
            n * (pos.0 - minx) as usize,
            n * (pos.1 - miny) as usize,
            1,
        );
    }
    res
}

fn find_pattern(img: &Tile, pattern: &Tile) -> Vec<(i64, i64)> {
    for flip_x in &[false, true] {
        for flip_y in &[false, true] {
            if *flip_x && *flip_y {
                continue;
            }
            for rotation in &Dir::ENUM {
                let tm = Transform {
                    rotation: rotation.clone(),
                    flip_x: *flip_x,
                    flip_y: *flip_y,
                };
                let img1 = img.transform(&tm);
                let matches = img1.find_matches(pattern);
                if matches.len() > 0 {
                    return matches;
                }
            }
        }
    }
    Vec::new()
}

pub(crate) fn solution() {
    let tiles: TileSet = fs::read_to_string(&common::data_file(20))
        .unwrap()
        .split("\n\n")
        .filter_map(|s| Tile::parse(s, true))
        .collect();
    let mapping = create_mapping(&tiles);
    let corners = find_corners(&tiles, &mapping);
    let res1: u64 = corners.iter().map(|i| tiles[*i].id).product();
    println!("Answer 1: {}", res1);

    let pattern = Tile::parse(
        &fs::read_to_string(&common::extra_file(20, "pattern")).unwrap(),
        false,
    )
    .unwrap();
    let layout = get_layout(&tiles, &mapping);
    let img = blit_layout(&tiles, &layout);
    let m = find_pattern(&img, &pattern);
    println!(
        "Answer 2: {}",
        img.count('#') - m.len() * pattern.count('#')
    );
}
